File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.py
def solution(players, callings):
    playerDictionary = {players[i] : i for i in range(len(players))}
    
    def passPlayer(name, playerDictionary, players):
        rightIndex = playerDictionary.get(name)
        leftIndex = rightIndex - 1
        leftName = players[leftIndex]
        players[leftIndex] = name
        players[rightIndex] = leftName
        playerDictionary[leftName] = rightIndex
        playerDictionary[name] = leftIndex
        return (playerDictionary, players)
    
    for name in callings:
        playerDictionary, players = passPlayer(name, playerDictionary, players)

    return players

File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.java
import java.util.HashMap;

class Solution {
    public String[] solution(String[] players, String[] callings) {
        HashMap<String, Integer> playerMap = new HashMap<>();
        for (int i = 0; i < players.length; i++) {
            playerMap.put(players[i], i);
        }
        for (String name : callings) {
            pass(name, playerMap, players);
        }
        return players;
    }
    public void pass(String name, HashMap<String, Integer> playerMap, String[] players){
        int right = playerMap.get(name);
        int left = right - 1;
        String leftName = players[left];
        playerMap.replace(leftName, right);
        playerMap.replace(name, left);
        players[left] = name;
        players[right] = leftName;
    }
}
